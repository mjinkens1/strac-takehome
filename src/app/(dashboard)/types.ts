export interface DriveFile {
  id: string;
  name: string;
  mimeType: string;
  modifiedTime: string;
}
