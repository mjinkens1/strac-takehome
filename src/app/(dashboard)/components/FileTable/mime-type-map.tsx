import {
  DocumentIcon,
  DocumentTextIcon,
  FolderIcon,
  PresentationChartBarIcon,
  TableCellsIcon,
} from '@heroicons/react/24/outline';
import { JSX } from 'react';

export const mimeTypeMap: Record<string, { label: string; icon: JSX.Element }> = {
  'application/pdf': {
    label: 'PDF Document',
    icon: <DocumentTextIcon className="mr-1 inline-block size-4 text-red-500" />,
  },
  'application/msword': {
    label: 'Word Document',
    icon: <DocumentIcon className="mr-1 inline-block size-4 text-blue-600" />,
  },
  'application/vnd.openxmlformats-officedocument.wordprocessingml.document': {
    label: 'Word Document',
    icon: <DocumentIcon className="mr-1 inline-block size-4 text-blue-600" />,
  },
  'application/vnd.ms-excel': {
    label: 'Excel Spreadsheet',
    icon: <TableCellsIcon className="mr-1 inline-block size-4 text-green-600" />,
  },
  'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet': {
    label: 'Excel Spreadsheet',
    icon: <TableCellsIcon className="mr-1 inline-block size-4 text-green-600" />,
  },
  'application/vnd.ms-powerpoint': {
    label: 'PowerPoint Presentation',
    icon: <PresentationChartBarIcon className="mr-1 inline-block size-4 text-orange-500" />,
  },
  'application/vnd.openxmlformats-officedocument.presentationml.presentation': {
    label: 'PowerPoint Presentation',
    icon: <PresentationChartBarIcon className="mr-1 inline-block size-4 text-orange-500" />,
  },
  'application/vnd.google-apps.document': {
    label: 'Google Docs',
    icon: <DocumentTextIcon className="mr-1 inline-block size-4 text-sky-500" />,
  },
  'application/vnd.google-apps.spreadsheet': {
    label: 'Google Sheets',
    icon: <TableCellsIcon className="mr-1 inline-block size-4 text-lime-600" />,
  },
  'application/vnd.google-apps.presentation': {
    label: 'Google Slides',
    icon: <PresentationChartBarIcon className="mr-1 inline-block size-4 text-yellow-500" />,
  },
  'application/vnd.google-apps.folder': {
    label: 'Google Drive Folder',
    icon: <FolderIcon className="mr-1 inline-block size-4 text-gray-500" />,
  },
  'image/jpeg': {
    label: 'JPEG Image',
    icon: <DocumentIcon className="mr-1 inline-block size-4 text-pink-400" />,
  },
  'image/png': {
    label: 'PNG Image',
    icon: <DocumentIcon className="mr-1 inline-block size-4 text-cyan-400" />,
  },
  'image/gif': {
    label: 'GIF Image',
    icon: <DocumentIcon className="mr-1 inline-block size-4 text-purple-500" />,
  },
  'text/plain': {
    label: 'Text File',
    icon: <DocumentTextIcon className="mr-1 inline-block size-4 text-gray-400" />,
  },
  'application/zip': {
    label: 'ZIP Archive',
    icon: <DocumentIcon className="mr-1 inline-block size-4 text-yellow-600" />,
  },
  'application/json': {
    label: 'JSON File',
    icon: <DocumentTextIcon className="mr-1 inline-block size-4 text-emerald-500" />,
  },
  'text/csv': {
    label: 'CSV File',
    icon: <TableCellsIcon className="mr-1 inline-block size-4 text-lime-600" />,
  },
  'image/heif': {
    label: 'HEIF Image',
    icon: <DocumentIcon className="mr-1 inline-block size-4 text-pink-400" />,
  },
  'video/mp4': {
    label: 'MP4 Video',
    icon: <DocumentIcon className="mr-1 inline-block size-4 text-violet-500" />,
  },
  'video/webm': {
    label: 'WebM Video',
    icon: <DocumentIcon className="mr-1 inline-block size-4 text-violet-500" />,
  },
  'video/quicktime': {
    label: 'MOV Video',
    icon: <DocumentIcon className="mr-1 inline-block size-4 text-violet-500" />,
  },
  'video/x-msvideo': {
    label: 'AVI Video',
    icon: <DocumentIcon className="mr-1 inline-block size-4 text-violet-500" />,
  },
};
