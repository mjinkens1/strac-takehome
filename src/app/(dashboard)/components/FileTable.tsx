import { useState, useCallback, memo, useMemo } from "react";
import { DriveFile } from "../types";
import { Tooltip } from "./Tooltip";
import {
  ArrowDownTrayIcon,
  CheckIcon,
  TrashIcon,
} from "@heroicons/react/24/outline";
import { mimeTypeMap } from "../mime-type-map";

interface FileTableProps {
  files: DriveFile[];
  onDelete: (fileId: string) => void;
  nextPageToken?: string;
  loadingMore: boolean;
  onLoadMore: (pageToken: string) => void;
}

export function FileTable({
  files,
  onDelete,
  nextPageToken,
  loadingMore,
  onLoadMore,
}: FileTableProps) {
  const sortedFies = useMemo(
    () =>
      [...files].sort(
        (a, b) =>
          new Date(b.modifiedTime).getTime() -
          new Date(a.modifiedTime).getTime()
      ),
    [files]
  );

  return (
    <div className="flex flex-col justify-between h-full rounded border border-[var(--color-card-border)]">
      <div className="relative min-h-0 overflow-y-auto rounded bg-[var(--color-card-bg)] shadow-xl backdrop-blur-md">
        <table className="h-full min-w-full table-auto text-sm">
          <thead className="sticky top-0 z-10 border-b border-[var(--color-card-border)] bg-[var(--color-card-bg)] text-left text-muted-foreground font-bold">
            <tr>
              <th className="px-6 py-4 w-[40%]">Filename</th>
              <th className="px-6 py-4 w-[25%]">Type</th>
              <th className="px-6 py-4 w-[25%]">Modified</th>
              <th className="px-6 py-4 w-[10%] text-right">Actions</th>
            </tr>
          </thead>
          <tbody className="relative">
            {sortedFies.map((file) => (
              <MemoizedFileRow key={file.id} file={file} onDelete={onDelete} />
            ))}

            {nextPageToken && (
              <tr>
                <td colSpan={4} className="px-6 py-4">
                  <button
                    onClick={() => onLoadMore(nextPageToken)}
                    className="w-full justify-center flex items-center text-blue-600 hover:text-blue-800"
                  >
                    {loadingMore ? (
                      <div className="size-5 border-2 border-blue-600 border-t-transparent rounded-full animate-spin" />
                    ) : (
                      "Load More"
                    )}
                  </button>
                </td>
              </tr>
            )}
          </tbody>
        </table>
      </div>
    </div>
  );
}

const FileRow = ({
  file,
  onDelete,
}: {
  file: DriveFile;
  onDelete: (id: string) => void;
}) => {
  const [progress, setProgress] = useState<number>(0);
  const [downloading, setDownloading] = useState(false);
  const [downloaded, setDownloaded] = useState(false);

  const handleDownload = useCallback(
    async (e: React.MouseEvent<HTMLAnchorElement>) => {
      e.preventDefault();
      setProgress(0);
      setDownloading(true);

      try {
        const res = await fetch(
          `/api/drive/download?id=${file.id}&name=${encodeURIComponent(
            file.name
          )}`
        );
        if (!res.body) throw new Error("No response body");

        const reader = res.body.getReader();
        const contentLength = +(res.headers.get("Content-Length") || 0);

        let received = 0;
        const chunks = [];

        while (true) {
          const { done, value } = await reader.read();
          if (done) break;
          if (value) {
            chunks.push(value);
            received += value.length;
            if (contentLength) {
              const pct = Math.round((received / contentLength) * 100);
              setProgress(pct);
            }
          }
        }

        const blob = new Blob(chunks);
        const url = window.URL.createObjectURL(blob);
        const a = document.createElement("a");
        a.href = url;
        a.download = file.name;
        document.body.appendChild(a);
        a.click();
        window.URL.revokeObjectURL(url);
        document.body.removeChild(a);

        setDownloaded(true);
      } catch (err) {
        console.error("Download error", err);
      } finally {
        setDownloading(false);
      }
    },
    [file]
  );

  return (
    <tr className="border-t border-[var(--color-card-border)] hover:bg-[var(--muted)]">
      <td className="w-[40%] font-medium text-[var(--foreground)]">
        <Tooltip text={file.name}>
          <div className="px-6 py-4 max-w-80 truncate">{file.name}</div>
        </Tooltip>
      </td>
      <td className="w-[25%] px-6 py-4">
        <div className="flex items-center gap-2">
          {mimeTypeMap[file.mimeType]?.icon}
          {mimeTypeMap[file.mimeType]?.label || file.mimeType}
        </div>
      </td>
      <td className="w-[25%] px-6 py-4">
        {new Date(file.modifiedTime).toLocaleString()}
      </td>
      <td className="w-[10%] px-6 py-4 text-right space-x-4 whitespace-nowrap">
        <a
          href={`/api/drive/download?id=${file.id}&name=${encodeURIComponent(
            file.name
          )}`}
          download={file.name}
          onClick={handleDownload}
          className={`text-blue-600 hover:text-blue-800 inline-block ${
            downloading || downloaded ? "pointer-events-none" : ""
          }`}
        >
          {downloaded ? (
            <CheckIcon className="size-5 text-green-500" />
          ) : downloading ? (
            <div className="relative size-5">
              <svg
                viewBox="0 0 36 36"
                className="absolute top-0 left-0 size-full"
              >
                <circle
                  cx="18"
                  cy="18"
                  r="16"
                  fill="none"
                  className="stroke-gray-300 dark:stroke-zinc-700"
                  strokeWidth="4"
                />
                <circle
                  cx="18"
                  cy="18"
                  r="16"
                  fill="none"
                  className="stroke-blue-600 transition-all duration-400"
                  strokeWidth="4"
                  strokeDasharray={2 * Math.PI * 16}
                  strokeDashoffset={2 * Math.PI * 16 * (1 - progress / 100)}
                  strokeLinecap="round"
                />
              </svg>
            </div>
          ) : (
            <Tooltip text="Download">
              <ArrowDownTrayIcon className="size-5" />
            </Tooltip>
          )}
        </a>

        <Tooltip text="Delete">
          <button
            onClick={() => onDelete(file.id)}
            className="text-red-500 hover:text-red-700"
          >
            <TrashIcon className="size-5" />
          </button>
        </Tooltip>
      </td>
    </tr>
  );
};

const MemoizedFileRow = memo(FileRow);
MemoizedFileRow.displayName = "FileRow";
