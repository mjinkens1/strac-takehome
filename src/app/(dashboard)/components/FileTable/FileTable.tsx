import { useMemo } from 'react';

import { DriveFile } from '../../types';
import { MemoizedFileRow } from './components/FileRow';
import { LoadMoreRow } from './components/LoadMoreRow';
import { TableHeader } from './components/TableHeader';

interface FileTableProps {
  files: DriveFile[];
  onDelete: (fileId: string) => void;
  nextPageToken?: string;
  loadingMore: boolean;
  onLoadMore: (pageToken: string) => void;
}

export function FileTable({
  files,
  onDelete,
  nextPageToken,
  loadingMore,
  onLoadMore,
}: FileTableProps) {
  const sortedFiles = useMemo(
    () =>
      [...files].sort(
        (a, b) => new Date(b.modifiedTime).getTime() - new Date(a.modifiedTime).getTime(),
      ),
    [files],
  );

  return (
    <div className="flex h-full flex-col justify-between rounded border border-[var(--color-card-border)]">
      <div className="relative min-h-0 overflow-y-auto rounded bg-[var(--color-card-bg)] shadow-xl backdrop-blur-md">
        <table className="h-full min-w-full table-auto text-sm">
          <TableHeader />
          <tbody className="relative">
            {sortedFiles.map((file) => (
              <MemoizedFileRow key={file.id} file={file} onDelete={onDelete} />
            ))}

            {nextPageToken && (
              <LoadMoreRow
                nextPageToken={nextPageToken}
                loadingMore={loadingMore}
                onLoadMore={onLoadMore}
              />
            )}
          </tbody>
        </table>
      </div>
    </div>
  );
}
