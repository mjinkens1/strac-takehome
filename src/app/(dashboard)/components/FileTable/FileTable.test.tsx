import { fireEvent, render, screen } from '@testing-library/react';

import { DriveFile } from '../../types';
import { FileTable } from './FileTable';

// Mock the child components
jest.mock('./components/FileRow', () => ({
  MemoizedFileRow: ({ file, onDelete }: { file: DriveFile; onDelete: (id: string) => void }) => (
    <tr data-testid={`file-row-${file.id}`}>
      <td>{file.name}</td>
      <td>
        <button onClick={() => onDelete(file.id)}>Delete</button>
      </td>
    </tr>
  ),
}));

jest.mock('./components/TableHeader', () => ({
  TableHeader: () => <thead data-testid="table-header" />,
}));

jest.mock('./components/LoadMoreRow', () => ({
  LoadMoreRow: ({
    onLoadMore,
    nextPageToken,
  }: {
    onLoadMore: (token: string) => void;
    nextPageToken: string;
  }) => (
    <tr>
      <td>
        <button data-testid="load-more" onClick={() => onLoadMore(nextPageToken)}>
          Load More
        </button>
      </td>
    </tr>
  ),
}));

describe('FileTable', () => {
  const mockFiles = [
    {
      id: '1',
      name: 'test1.pdf',
      modifiedTime: '2024-01-01T00:00:00Z',
      mimeType: 'application/pdf',
    },
    {
      id: '2',
      name: 'test2.pdf',
      modifiedTime: '2024-01-02T00:00:00Z',
      mimeType: 'application/pdf',
    },
  ];

  const mockProps = {
    files: mockFiles,
    onDelete: jest.fn(),
    loadingMore: false,
    onLoadMore: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders table with header and file rows', () => {
    render(<FileTable {...mockProps} />);

    expect(screen.getByTestId('table-header')).toBeInTheDocument();
    expect(screen.getByTestId('file-row-1')).toBeInTheDocument();
    expect(screen.getByTestId('file-row-2')).toBeInTheDocument();
  });

  it('sorts files by modified time in descending order', () => {
    render(<FileTable {...mockProps} />);

    const rows = screen.getAllByTestId(/file-row/);
    // Most recent file should be first
    expect(rows[0]).toHaveAttribute('data-testid', 'file-row-2');
    expect(rows[1]).toHaveAttribute('data-testid', 'file-row-1');
  });

  it('shows load more button when nextPageToken is provided', () => {
    render(<FileTable {...mockProps} nextPageToken="next-token" />);

    const loadMoreButton = screen.getByTestId('load-more');
    expect(loadMoreButton).toBeInTheDocument();

    fireEvent.click(loadMoreButton);
    expect(mockProps.onLoadMore).toHaveBeenCalledWith('next-token');
  });

  it('handles file deletion', () => {
    render(<FileTable {...mockProps} />);

    const deleteButtons = screen.getAllByText('Delete');
    fireEvent.click(deleteButtons[0]);

    expect(mockProps.onDelete).toHaveBeenCalledWith(mockFiles[1].id);
  });

  it('matches snapshot', () => {
    const { container } = render(<FileTable {...mockProps} />);
    expect(container).toMatchSnapshot();
  });
});
