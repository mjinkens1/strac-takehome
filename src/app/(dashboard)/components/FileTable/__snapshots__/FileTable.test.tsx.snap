// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileTable matches snapshot 1`] = `
<div>
  <div
    class="flex flex-col justify-between h-full rounded border border-[var(--color-card-border)]"
  >
    <div
      class="relative min-h-0 overflow-y-auto rounded bg-[var(--color-card-bg)] shadow-xl backdrop-blur-md"
    >
      <table
        class="h-full min-w-full table-auto text-sm"
      >
        <thead
          data-testid="table-header"
        />
        <tbody
          class="relative"
        >
          <tr
            data-testid="file-row-2"
          >
            <td>
              test2.pdf
            </td>
            <td>
              <button>
                Delete
              </button>
            </td>
          </tr>
          <tr
            data-testid="file-row-1"
          >
            <td>
              test1.pdf
            </td>
            <td>
              <button>
                Delete
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
